src/profile/ProfileForm.tsx

```diff
@@ -37,585 +23,315 @@ interface ChangedUser {
+      first_name: props.user?.first_name,
+      last_name: props.user?.last_name,
+      email: props.user?.email,
+      is_editor: props.user?.is_editor || false,
+      writer_name: props.user?.writer_name,
+    });
+  }, [props.user, reset]);
+
+  // manually register the is_editor field since we handle it with a select
+  useEffect(() => {
+    register({
+      name: "is_editor",
+    });
+  }, [register]);
+
+  const [showPassword, setShowPassword] = useState<boolean>(false);
+
+  const submit = async (vals: ProfileFormVals) => {
+    const cleaned = cleanFormData(vals);
+
+    const editingMe = props.user?.username === auth.user?.username;
+    try {
+      if (props.user === undefined) {
+        await auth.post<ProfileFormVals>("users/", cleaned);
+      } else {
+        await auth.patch<ProfileFormVals>(
+          editingMe ? "me/" : `users/${vals?.username}/`,
+          cleaned,
+          editingMe
+        );
       }
+    } catch (err) {
+      const apiErrors = err as UserAPIError;
+      props.context.setError(
+        apiErrors.cur_password?.map((error) => ({
+          name: "cur_password",
+          type: "server_error",
+          message: error,
+        }))
+      );
     }
-  };
 
-  const onSubmit = (evt?: React.FormEvent<HTMLFormElement>) => {
-    evt?.preventDefault();
-
-    if (
-      (Object.keys(state.changedUser) as Array<keyof ChangedUser>).every(
-        (k) =>
-          state.changedUser[k] === undefined ||
-          (k === "is_editor" && state.changedUser[k] === user?.is_editor)
-      )
-    ) {
-      return Promise.resolve();
+    if (props.onSubmit) {
+      await props.onSubmit(vals);
     }
-
-    if (Object.values(state.errors).flat().length > 1) {
-      dispatch({
-        type: "set general error",
-        errors: [ERRORS.FORMS.NOT_YET_VALID],
-      });
-      return Promise.reject();
-    }
-
-    dispatch({ type: "is loading" });
-
-    return (user === undefined
-      ? auth.post<ChangedUser>("users/", state.changedUser)
-      : auth.patch<ChangedUser>(
-          user === auth.user ? "me/" : `users/${user?.username}/`,
-          state.changedUser,
-          user === auth.user
-        )
-    ).then(
-      () => {
-        dispatch({ type: "done loading success" });
-      },
-      (err: UserAPIError | string) => {
-        dispatch({ type: "done loading error", errors: err });
-      }
-    );
   };
 
-  useImperativeHandle(ref, () => ({
-    submit: onSubmit,
-  }));
-
   return (
-    <Form noValidate onSubmit={onSubmit}>
-      {state.generalErrors.length > 0 &&
-        state.generalErrors.map((err) => (
-          <Alert key={err} variant="danger">
-            {ERRORS[err]}
-          </Alert>
-        ))}
-
-      {state.successMessage.length > 0 && (
-        <Alert
-          variant="success"
-          onClose={() => dispatch({ type: "set success message", message: "" })}
-          dismissible
-        >
-          {state.successMessage}
-        </Alert>
-      )}
```

Comment:
These alerts should be kept in, so that the user knows if submission was successful or not.

Reply:
https://github.com/UWmathNEWS/slugline-web/pull/13/commits/e85313c02adcbb5620a46d8d78d4a455ce6c3f19